pkg/certificate/mrc_reconciler.go

```diff
@@ -1,214 +1,228 @@
+	_, ok := validMRCIntents[intent]
+	if !ok {
+		log.Error().Err(ErrUnexpectedMRCIntent).Msgf("unable to find %s intent in set of valid intents", intent)
+		return ErrUnexpectedMRCIntent
 	}
-)
 
-func (uc UseCase) String() string {
-	return string(uc)
-}
-func (m *Manager) handleMRCEvent(event MRCEvent) error {
-	mrc := event.MRC
-	// TODO(5046): remove this first call to setIssuers.
-	err := m.setIssuers(event.MRC)
-	if err == nil {
-		return nil
-	}
-	log.Err(err).Msg("error setting issuers on mrc")
-	// TODO(5046): improve logging in this method.
-	if shouldUpdateMRCComponentStatus(mrc) {
-		if err := m.updateMRCComponentStatus(mrc); err != nil {
-			return err
-		}
-	} else if shouldUpdateState(mrc) {
-		if err := m.updateMRCState(mrc); err != nil {
-			return err
-		}
-	} else if m.shouldSetIssuers(mrc) {
-		log.Info().Msgf("setting new certificate issuers on the Certificate Manager for MRC %s in stage %s", mrc.GetName(), mrc.Status.State)
-		return m.setIssuers(mrc)
-	}
+	log.Debug().Msgf("validated MRC intent %s", intent)
 	return nil
 }
 
-// Component statuses determine if we're ready to transition the mrc to the next state.
-// TODO(5046): add unit tests for this, and other methods in here.
-func shouldUpdateMRCComponentStatus(mrc *v1alpha2.MeshRootCertificate) bool {
-	if isTerminal(mrc) {
-		return false
+// getSigningAndValidatingMRCs returns the signing and validating MRCs from a list of MRCs
+func getSigningAndValidatingMRCs(mrcList []*v1alpha2.MeshRootCertificate) (*v1alpha2.MeshRootCertificate, *v1alpha2.MeshRootCertificate, error) {
+	if len(mrcList) == 0 {
+		log.Error().Err(ErrNoMRCsFound).Msg("when handling MRC event, found no MRCs in OSM control plane namespace")
+		return nil, nil, ErrNoMRCsFound
+	}
+	if len(mrcList) > 2 {
+		log.Error().Err(ErrNumMRCExceedsMaxSupported).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrNumMRCExceedsMaxSupported)).
+			Msgf("expected 2 or less MRCs in the OSM control plane namespace, found %d", len(mrcList))
+		return nil, nil, ErrNumMRCExceedsMaxSupported
 	}
 
-	if len(mrc.Status.Conditions) == 0 {
-		return false
+	var mrc1, mrc2 *v1alpha2.MeshRootCertificate
+	mrc1 = mrcList[0]
+	if len(mrcList) == 2 {
+		mrc2 = mrcList[1]
+	} else {
+		log.Trace().Msgf("found single MRC in the mesh when handling MRC event for MRC %s", mrc1.Name)
+		// if there is only one MRC, set mrc2 equal to mrc1
+		mrc2 = mrc1
 	}
 
-	transitionAfter := mrc.Status.TransitionAfter
-	if transitionAfter == nil {
-		return false
+	if mrc1 == nil || mrc2 == nil {
+		log.Error().Err(ErrUnexpectedNilMRC).Msg("unexpected nil MRC provided when validating MRC intents")
+		return nil, nil, ErrUnexpectedNilMRC
 	}
 
-	//The truncate is needed so we truncate the monotonic clock.
-	return transitionAfter.Truncate(0).After(time.Now())
-}
+	intent1 := mrc1.Spec.Intent
+	intent2 := mrc2.Spec.Intent
+
+	if err := validateMRCIntent(intent1); err != nil {
```

Comment:
should these be in a validating webhook?

Reply:
Yes, #5205 will add this validation to a validating webhook.